### PR TITLE
feat(open-pr-comments): javascript parser

### PR DESCRIPTION
**Pull request by cathteng**
Adds a Javascript parser for Open PR Comments. This means we will be parsing the git hunk headers which look like this for Python `@@ -20,3 +20,254 @@ def test_python_in_class(self):` in order to find functions being modified in Github PR files.

The Javascript parser can find functions declared in 4 ways using regexes.

- Function declaration e.g. `function hello(argument1, argument2)`
- Arrow function e.g. `export const blue = (argument) => {`
- Function expression e.g. `const planet = async function(argument) {`
- Function constructor e.g. `const constructor = new Function(`

In our React Typescript codebase we do have instances of functions like [`renderGroupInfo()`](https://github.com/getsentry/sentry/pull/34358/files#diff-de1f35912899e804d595f4ea3c2050f573399a77dac1ec994022ecbfc708ac39L98) inside a class, but they don't seem to be picked up by git.

For [Open PR Comments for Javascript](https://github.com/getsentry/team-enterprise/issues/41)

### CODE REVIEW

File: src/sentry/tasks/integrations/github/patch_parsers.py [hunk 1]
```diff
@@ -31,4 +31,33 @@ def         - Matches the literal characters "def".
...
+        ]
+
+        functions = set()
+        for regex in regexes:
+            functions.update(set(re.findall(regex, patch, flags=re.M)))
+
+        return functions
+
+
 PATCH_PARSERS = {"py": PythonParser}
```
Comment (cathteng) [inline src/sentry/tasks/integrations/github/patch_parsers.py:63]: TODO: add javascript here. `js` will definitely be supported, and i think `jsx` can be too

File: tests/sentry/tasks/integrations/github/test_patch_parsers.py [hunk 2]
```diff
@@ -20,3 +20,254 @@ def test_python_in_class(self):
             "delete",
             "put",
         }
+
+    def test_javascript_simple(self):
+        patch = """@@44,38@@ function hello(argument1, argument2)
+
+@@44,38@@ export default function world({argument}) {
```
Comment (billyvg) [inline tests/sentry/tasks/integrations/github/test_patch_parsers.py:27]: Thinking about other test cases here...

Async? What about no args? What if `{` is on next line

```suggestion
export default async function world({argument}) {
```

File: tests/sentry/tasks/integrations/github/test_patch_parsers.py [hunk 3]
```diff
@@ -20,3 +20,254 @@ def test_python_in_class(self):
...
+
+@@44,38@@ export const blue = (argument) => {
```
Comment (billyvg) [inline tests/sentry/tasks/integrations/github/test_patch_parsers.py:29]: Arrow functions don't always have a curly brace

File: tests/sentry/tasks/integrations/github/test_patch_parsers.py (hunk 2 above)
Comment (cathteng) [reply tests/sentry/tasks/integrations/github/test_patch_parsers.py:27]: the `@@44, 38@@` is how git represents the hunk header with the line number. anything else is code itself

the regex accounts for any words before the word `function`, but i can add more

i can add a case for { on the next line

File: tests/sentry/tasks/integrations/github/test_patch_parsers.py (hunk 3 above)
Comment (cathteng) [reply tests/sentry/tasks/integrations/github/test_patch_parsers.py:29]: added a case

File: tests/sentry/tasks/integrations/github/test_patch_parsers.py (hunk 2 above)
Comment (billyvg) [reply tests/sentry/tasks/integrations/github/test_patch_parsers.py:27]: Oops sorry, wasn't meant to be a github suggestion

File: tests/sentry/tasks/integrations/github/test_patch_parsers.py (hunk 3 above)
Comment (markstory) [reply tests/sentry/tasks/integrations/github/test_patch_parsers.py:31]: You can also have arrow functions like `const doThing = arg => arg + 1`

File: src/sentry/tasks/integrations/github/language_parsers.py [hunk 4]
```diff
@@ -96,4 +96,39 @@ def generate_function_name_conditions(function_names: List[str], stack_frame: in
         )
 
 
+class JavascriptParser(LanguageParser):
+    @staticmethod
+    def extract_functions_from_patch(patch: str) -> Set[str]:
+        r"""
+        Type of function declaration    Example
+        Function declaration:           function hello(argument1, argument2)
+        Arrow function:                 export const blue = (argument) => {
```
Comment (sentaur-athena) [inline src/sentry/tasks/integrations/github/language_parsers.py:105]: The word export here doesn't really matter, right?

File: tests/sentry/tasks/integrations/github/test_language_parsers.py
Comment (schew2381) [inline tests/sentry/tasks/integrations/github/test_language_parsers.py:1]: Might be good to split this test file by language

**Review by sentaur-athena — APPROVED** [review]
The regex and tests look good. I don't know all the js patterns though so I can't say if the regex would find things that are not functions. We'll see than while testing in prod 😝

File: src/sentry/tasks/integrations/github/language_parsers.py (hunk 4 above)
Comment (cathteng) [reply src/sentry/tasks/integrations/github/language_parsers.py:105]: it doesn't matter, but i included it as an example of some words in front of the `constant` that we also handle

**Pull request merged**